pytket/pytket/mbqc_compile/mpattern.py

```diff
@@ -0,0 +1,1074 @@
+        or by non-Clifford gates and turning each segment to a ZX diagram. Then, if no
+        architecture is given the ciruits are joined up in the fully connected setting,
+        otherwise, they are routed on the architecture using one of two possible
+        strategies. For each strategy ("Separate" or "Sequential") call the 
+        corresponding method. Finally returns a tuple containing the routed circuit
+        and the dictionary mapping the inputs/outputs to the original circuit.
+        
+        :param arch:       The architecture to route onto ("None" treated as "FullyConnected")
+        :param type:       Architecture
+        
+        :param n:          Number of segments to attempt to split into.
+        :param type:       int
+        
+        :param splitStrat: Splitting strategy either by "Depth" or by "Gates".
+        :param type:       str
+        
+        :param routeStrat: Routing strategy either by "Separate" or "Sequential".
+        :param type:       str
+        
+        :returns:          A tuple containing a circuit and an i/o map dictionary.
+        :rtype:            tuple(Circuit,dict)
+        """
+        if (type(arch)==type(FullyConnected(0))) or (arch == None):
+            return self.unrouted_conversion(n,splitStrat)
+        else:
+            pattern_list = self.multi_conversion(n, splitStrat)
+            if routeStrat == "Separate":
+                return self.routed_conversion_separate(pattern_list,arch)
+            elif routeStrat == "Sequential":
+                return self.routed_conversion_sequential(pattern_list,arch)
+        
+    def routed_conversion_separate(self, pattern_list: list, arch: Architecture = None) -> tuple:
+        """
+        This method is given a list of tuples and an architecture as an input.
+        Each tuple contains a circuit and a dictionary mapping its inputs and outputs
+        to some original circuit qubits. This method routes each of the segments separately,
+        onto some given device architecture and then generates a network of SWAP gates
+        which also respect the architecture, to join each segment to the next one.
+        Finally returns a single tuple containing the resulting Circuit object and
+        the dictionary of the final i/o map.
+        
+        :param pattern_list: A list of tuples containing circuits and i/o maps.
+        :param type:         list(tuple(Circuit,dict))
+        
+        :param arch:         The architecture to route onto.
+        :param type:         Architecture
+        
+        :returns:            A tuple containing a circuit and an i/o map dictionary.
+        :rtype:              tuple(Circuit,dict)
+        """        
+        new_c = Circuit()
+        for q in arch.nodes:
+            new_c.add_qubit(q)
+        for p in range(len(pattern_list)):
+            (curr_circ,curr_map) = pattern_list[p]
+            route_map = {}
+            if p==0:
```

Comment:
can work required in the. p==0 case be done outside for for p in range(...) loop? to avoid this if else case